File: lib-segment/src/main/java/websiteschema/mpsegment/tools/accurary/SegmentAccuracy.java
package websiteschema.mpsegment.tools.accurary;

import websiteschema.mpsegment.core.SegmentResult;
import websiteschema.mpsegment.core.SegmentWorker;
import websiteschema.mpsegment.core.WordAtom;
import websiteschema.mpsegment.tools.PFRCorpusLoader;
import websiteschema.mpsegment.util.NumberUtil;
import websiteschema.mpsegment.util.StringUtil;

import java.io.IOException;
import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.Map;

public class SegmentAccuracy {

    private PFRCorpusLoader loader;
    private int totalWords = 0;
    private int correct = 0;
    private int wrong = 0;
    private double accuracyRate;

    private HashMap<String, Integer> allWordsAndFreqInCorpus = new HashMap<String, Integer>();
    private SegmentWorker segmentWorker;
    private Map<SegmentErrorType, ErrorAnalyzer> allErrorAnalyzer;

    public SegmentAccuracy(String testCorpus, SegmentWorker segmentWorker) throws IOException {
        this.segmentWorker = segmentWorker;
        initialErrorAnalyzer();
        loader = new PFRCorpusLoader(getClass().getClassLoader().getResourceAsStream(testCorpus));
    }

    public double getAccuracyRate() {
        return accuracyRate;
    }

    public int getWrong() {
        return wrong;
    }

    public int getTotalWords() {
        return totalWords;
    }

    public ErrorAnalyzer getErrorAnalyzer(SegmentErrorType errorType) {
        return allErrorAnalyzer.get(errorType);
    }

    public void checkSegmentAccuracy() {
        boolean isUseContextFreq = segmentWorker.isUseContextFreqSegment();
        segmentWorker.setUseContextFreqSegment(true);
        try {
            SegmentResult expectResult = loader.readLine();
            while (expectResult != null) {
                String sentence = expectResult.toOriginalString();
                SegmentResult actualResult = segmentWorker.segment(sentence);

                totalWords += expectResult.length();

                compare(expectResult, actualResult);
                expectResult = loader.readLine();
            }
        } catch (IOException ex) {
            ex.printStackTrace();
        } finally {
            postAnalysis();
            segmentWorker.setUseContextFreqSegment(isUseContextFreq);
        }
        assert (correct > 0 && totalWords > 0);
        accuracyRate = (double) correct / (double) totalWords;
    }

    private void initialErrorAnalyzer() {
        allErrorAnalyzer = new LinkedHashMap<SegmentErrorType, ErrorAnalyzer>();
        allErrorAnalyzer.put(SegmentErrorType.NER_NR, new NerNameErrorAnalyzer());
        allErrorAnalyzer.put(SegmentErrorType.NER_NS, new NerPlaceErrorAnalyzer());
        allErrorAnalyzer.put(SegmentErrorType.UnknownWord, new NewWordErrorAnalyzer());
        allErrorAnalyzer.put(SegmentErrorType.ContainDisambiguate, new ContainErrorAnalyzer());
        allErrorAnalyzer.put(SegmentErrorType.Other, new OtherErrorAnalyzer());
    }

    private void compare(SegmentResult expectResult, SegmentResult actualResult) {
        int lastMatchIndex = -1;
        for (int i = 0; i < expectResult.length(); i++) {
            WordAtom expectWord = expectResult.getWordAtom(i);
            recordWordFreqInCorpus(expectWord);
            int indexInOriginalString = expectResult.getWordIndexInOriginalString(i);
            int match = lookupMatch(actualResult, expectWord, lastMatchIndex + 1, indexInOriginalString);
            if (match >= 0) {
                lastMatchIndex = match;
                correct++;
            } else {
                wrong++;
            }
        }
    }

    private void recordWordFreqInCorpus(WordAtom word) {
        int freq = allWordsAndFreqInCorpus.containsKey(word.word) ? allWordsAndFreqInCorpus.get(word.word) + 1 : 1;
        allWordsAndFreqInCorpus.put(word.word, freq);
    }

    private int lookupMatch(SegmentResult actualResult, WordAtom expectWord, int start, final int indexInOriginalString) {
        for (int i = start; i < actualResult.length(); i++) {
            String actualWord = actualResult.getWord(i);
            if (isSameWord(expectWord.word, actualWord)) {
                if (actualResult.getWordIndexInOriginalString(i) == indexInOriginalString) {
                    return i;
                }
            }
        }
        analyzeErrorReason(actualResult, expectWord, start, indexInOriginalString);
        return -1;
    }

    private void analyzeErrorReason(SegmentResult actualResult, WordAtom expect, int start, int from) {
        String possibleErrorWord = lookupErrorWord(actualResult, expect, start, from);
        analyzeReason(expect, possibleErrorWord);
    }

    private void analyzeReason(WordAtom expect, String possibleErrorWord) {
        for(SegmentErrorType errorType : allErrorAnalyzer.keySet()) {
            ErrorAnalyzer analyzer = allErrorAnalyzer.get(errorType);
            boolean isErrorWord = analyzer.analysis(expect, possibleErrorWord);
            if(isErrorWord) {
                break;
            }
        }
    }

    private String lookupErrorWord(SegmentResult actualResult, WordAtom expect, int start, int from) {
        int to = from + expect.length();
        StringBuilder stringBuilder = new StringBuilder();
        for (int i = start; i < actualResult.length(); i++) {
            int indexInOriginalString = actualResult.getWordIndexInOriginalString(i);
            if (indexInOriginalString >= from && indexInOriginalString < to) {
                stringBuilder.append(actualResult.getWord(i)).append(" ");
            }
        }

        return stringBuilder.toString().trim();
    }

    private boolean isSameWord(String expect, String actual) {
        String expectWord = StringUtil.doUpperCaseAndHalfShape(expect);
        if (expectWord.equalsIgnoreCase(actual)) {
            return true;
        }
        if (Character.isDigit(actual.charAt(0))) {
            String number = NumberUtil.chineseToEnglishNumberStr(expect);
            if (actual.equals(number)) {
                return true;
            }
        }
        return false;
    }

    private void postAnalysis() {
        for (SegmentErrorType errorType : allErrorAnalyzer.keySet()) {
            getErrorAnalyzer(errorType).postAnalysis(allWordsAndFreqInCorpus);
        }
    }
}

File: lib-segment-scala/src/main/scala/websiteschema/mpsegment/tools/accurary/SegmentAccuracy.scala
package websiteschema.mpsegment.tools.accurary

import websiteschema.mpsegment.core.{SegmentResult, SegmentWorker, Word}
import websiteschema.mpsegment.dict.POSUtil
import websiteschema.mpsegment.tools.PFRCorpusLoader
import websiteschema.mpsegment.tools.accurary.SegmentErrorType._

import scala.collection.mutable._

class SegmentAccuracy(testCorpus: String, segmentWorker: SegmentWorker) extends SegmentResultCompareHook {

  private var loader: PFRCorpusLoader = null
  private var totalWords: Int = 0
  private var correct: Int = 0
  private var wrong: Int = 0
  private var accuracyRate: Double = 0D

  private val allWordsAndFreqInCorpus = HashMap[String, Int]()
  private var allErrorAnalyzer = new LinkedHashMap[SegmentErrorType, ErrorAnalyzer]()


  initialErrorAnalyzer()
  loader = PFRCorpusLoader(getClass().getClassLoader().getResourceAsStream(testCorpus))

  def getAccuracyRate() = accuracyRate

  def getWrong() = wrong

  def getTotalWords() = totalWords

  def getErrorAnalyzer(errorType: SegmentErrorType) = allErrorAnalyzer(errorType)

  def checkSegmentAccuracy() {
    val isUseContextFreq = segmentWorker.isUseContextFreqSegment()
    segmentWorker.setUseContextFreqSegment(true)
    try {
      var expectResult = loader.readLine()
      while (expectResult != null) {
        val sentence = expectResult.toOriginalString()
        val actualResult = segmentWorker.segment(sentence)

        NerNameStatisticData.scanNameWordCount(expectResult)
        NerNameStatisticData.scanRecognizedNameWordCount(actualResult)
//        println(actualResult)
        totalWords += expectResult.length()

        recordWordFreqInCorpus(expectResult)

        val comparator = new SegmentResultComparator(this)
        comparator.compare(expectResult, actualResult)
        expectResult = loader.readLine()
      }
    } catch {
      case ex: Throwable =>
        ex.printStackTrace()
    } finally {
      postAnalysis()
      segmentWorker.setUseContextFreqSegment(isUseContextFreq)
    }
    assert(correct > 0 && totalWords > 0)
    accuracyRate = correct.toDouble / totalWords.toDouble
  }

  private def initialErrorAnalyzer() {
    allErrorAnalyzer.put(SegmentErrorType.NER_NR, new NerNameErrorAnalyzer())
    allErrorAnalyzer.put(SegmentErrorType.NER_NS, new NerPlaceErrorAnalyzer())
    allErrorAnalyzer.put(SegmentErrorType.UnknownWord, new NewWordErrorAnalyzer())
    allErrorAnalyzer.put(SegmentErrorType.ContainDisambiguate, new ContainErrorAnalyzer())
    allErrorAnalyzer.put(SegmentErrorType.Other, new OtherErrorAnalyzer())
  }

  private def recordWordFreqInCorpus(expectResult: SegmentResult) {
    expectResult.foreach(word => {
      val freq = if (allWordsAndFreqInCorpus.contains(word.name)) allWordsAndFreqInCorpus(word.name) + 1 else 1
      allWordsAndFreqInCorpus.put(word.name, freq)
    })
  }

  private def postAnalysis() {
    for (errorType <- allErrorAnalyzer.keys) {
      getErrorAnalyzer(errorType).postAnalysis(allWordsAndFreqInCorpus)
    }
  }

  override def errorWordHook {
    wrong += 1
  }

  override def correctWordHook(expectWord: Word, matchedWord: Word, expectWordIndex: Int, matchedWordIndex: Int) {
    segmentCorrect(expectWord, matchedWord)
    correct += 1
  }

  private def segmentCorrect(expect: Word, actual: Word) {
    if (actual.pos == POSUtil.POS_NR) {
      NerNameStatisticData.correctRecognizedNameCount += 1
    }
  }

  override def analyzeReason(expect: Word, possibleErrorWord: String) {
    for (errorType <- allErrorAnalyzer.keys) {
      val analyzer = allErrorAnalyzer(errorType)
      val isErrorWord = analyzer.analysis(expect, possibleErrorWord)
      if (isErrorWord) {
        return
      }
    }
  }
}





